File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/OtherWeaponTraitEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum OtherWeaponTraitEnum implements WeaponTraitEnum {
    AGILE("agile"),
    FINESSE("finesse"),
    FREE_HAND("freeHand"),
    SHOVE("shove"),
    REACH("reach"),
    TRIP("trip"),
    PARRY("parry"),
    SWEEP("sweep"),
    FORCEFUL("forceful"),
    BACKSWING("backswing"),
    NONLETHAL("nonlethal"),
    ATTACHED_TO_SHIELD("attachedToShield"),
    BACKSTABBER("backstabber"),
    TWIN("twin");

    private final String value;
    private final static Map<String, OtherWeaponTraitEnum> CONSTANTS = new HashMap<>();

    static {
        for (OtherWeaponTraitEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    OtherWeaponTraitEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static OtherWeaponTraitEnum fromValue(String value) {
        OtherWeaponTraitEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/OtherWeaponTraitEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class OtherWeaponTraitEnum(private val value: String) : WeaponTraitEnum {
    AGILE("agile"), FINESSE("finesse"), FREE_HAND("freeHand"), SHOVE("shove"), REACH("reach"), TRIP("trip"), PARRY("parry"), SWEEP("sweep"), FORCEFUL("forceful"), BACKSWING("backswing"), NONLETHAL("nonlethal"), ATTACHED_TO_SHIELD("attachedToShield"), BACKSTABBER("backstabber"), TWIN("twin");

    companion object {
        private val CONSTANTS: MutableMap<String, OtherWeaponTraitEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): OtherWeaponTraitEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}